### PR TITLE
Add LSF support

### PR DESCRIPTION
**Pull request by ajtritt**
## What does this PR do?

Replaces #4966 

Add support for running PyTorch Lightning on systems managed by LSF.

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x]  Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._

## Did you have fun?
Always.

### CODE REVIEW

**Comment by ajtritt** [comment]
This PR probably requires some changes/additions to the docs. If everyone is okay with these changes I will update the documentation.

File: pytorch_lightning/cluster_environments/lsf_environment.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+import os
+import re
+import socket
+import warnings
+from pytorch_lightning import _logger as log
+from pytorch_lightning.cluster_environments.cluster_environment import ClusterEnvironment
+
+
+class LSFEnvironment(ClusterEnvironment):
```
Comment (tchaton) [inline pytorch_lightning/cluster_environments/lsf_environment.py:23]: From your knowledge, do you think it would be worth to add ci / cd for the LSF ClusterEnvironment. It seemed possible using https://github.com/IBMSpectrumComputing/lsf-hybrid-cloud/tree/master/LSF_On_AWS.

Comment (ajtritt) [reply pytorch_lightning/cluster_environments/lsf_environment.py:23]: Being able to test this would be nice, but I'm not aware of any CI service that uses LSF. 

That is for extending an existing LSF system to use cloud resources, so I don't think it would work as a CI/CD solution for us.

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -434,6 +434,15 @@ def fit(
         # bookkeeping
         self._state = TrainerState.RUNNING
 
+        # ----------------------------
+        # SET UP TRAINING
+        # ----------------------------
+        self.accelerator_backend = self.accelerator_connector.select_accelerator()
+        self.local_rank = self.accelerator_backend.cluster_environment.local_rank()
+        self.node_rank = self.accelerator_backend.cluster_environment.node_rank()
+        self.global_rank = self.accelerator_backend.cluster_environment.global_rank()
```
[the remark this answers is not in the document]
Comment (ajtritt) [reply pytorch_lightning/trainer/trainer.py:443]: I don't see that method. I assume you're suggesting I add that? 

There seems to be a lot of bidirectional references for the sake of altering the Trainer state with single method calls. Personally, I found it very confusing and it made it really difficult to diagnose problems. If it's a design choice the PL team has made, I respect that and will heed your suggestions, but I thought I would offer up my experience.

**Comment by carmocca** [comment]
Hi @ajtritt 

There are many changes in this PR which are not strictly necessary to "add LSF support". Can we split this PR into two? One for refactors/improvements and the other just for LSF support?

Thanks for the work you put into this! We really appreciate it.

**Comment by bbelgodere** [comment]
@ajtritt I'm hoping you may have some insight on this. I'm trying to get PL to work on our own LSF cluster, I can get ddp working with single node multi GPU, multi node single gpu, but dies with NCCL errors on multi node multi GPU. I've tried both ddp and ddp_spawn. To minimize any possible config/hardware drift in the cluster I'm using the same two nodes that I was able to get working with multi node single gpu. These are whitebox version of the V10032GB DGX with dual 100GB EDR connections to a single director class switch. 

* CUDA:
        - GPU:
                - Tesla V100-SXM2-32GB
                - Tesla V100-SXM2-32GB
                - Tesla V100-SXM2-32GB
                - Tesla V100-SXM2-32GB
        - available:         True
        - version:           11.0
* Packages:
        - numpy:             1.17.0
        - pyTorch_debug:     False
        - pyTorch_version:   1.7.1
        - pytorch-lightning: 1.1.6
        - tqdm:              4.49.0
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                -
        - processor:         x86_64
        - python:            3.7.9
        - version:           #1 SMP Fri Mar 27 14:35:58 UTC 2020
        
the other interesting thing I did find is that I can't get the exact same code to work on the exact same nodes, when interactive jobs have 2 GPUs available, even though gpus flag is set to 1 , if there is only 1 GPU available in the session it works.

File: pytorch_lightning/cluster_environments/lsf_environment.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+
+    This plugin expects the following environment variables:
+
+    LSB_JOBID
+      The LSF assigned job ID
+
+    LSB_HOSTS
+      The hosts used in the job. This string is expected to have the format "batch <rank_0_host> ...."
+
+    JSM_NAMESPACE_LOCAL_RANK
```
Comment (TheWitness) [inline pytorch_lightning/cluster_environments/lsf_environment.py:37]: This is a Coral based LSF integration and not a native LSF integration.  So, we need to be clear that LSF does not include JSM today.  To create a native LSF integration, this adaptation will have to change.  It may not be a major change, but someone will have to issue a follow-on integration to make that happen.

File: pytorch_lightning/accelerators/accelerator_connector.py [hunk 4]
```diff
@@ -182,84 +169,14 @@ def select_accelerator(self):
...
 
-        # choose the appropriate accelerator backend
         if self.trainer.use_ddp2:
             accelerator_backend = accelerators.DDP2Accelerator(
                 self.trainer,
                 cluster_env,
                 self.trainer.plugin_connector.ddp_plugin
             )
-
-        elif use_ddp_cpu_slurm:
```
Comment (Borda) [inline pytorch_lightning/accelerators/accelerator_connector.py:212]: is this also related to the LFS support?

Comment (ajtritt) [reply pytorch_lightning/accelerators/accelerator_connector.py:212]: No, just cleaning up the logic used to determine which backend to use.

Comment (Borda) [reply pytorch_lightning/accelerators/accelerator_connector.py:212]: sounds good, just please do this cleaning in separate PR :]

**Comment by edenlightning** [comment]
mind looking at failed tests?

**Comment by Borda** [comment]
@ajtritt there won't be any more releases in 1.1.x so I have changed your destination branch to feat 1.2
pls, there is no need to close this PR, just rebase on the target branch, or maybe easier just cherry-pick the commit (@kaushikb11 will help you if needed)

File: pytorch_lightning/cluster_environments/lsf_environment.py (hunk 3 above)
Comment (ajtritt) [reply pytorch_lightning/cluster_environments/lsf_environment.py:37]: What is Coral?  I determined these variables by printing the environment from a job i.e. they were not documented anywhere. Are you aware of any environment variables set by LSF that gives rank and size? I've only been able to find info [here](https://www.ibm.com/support/knowledgecenter/SSWRJV_10.1.0/lsf_config_ref/lsf_envars_job_exec.html)

**Comment by carmocca** [comment]
Hi @ajtritt

A large accelerator refactor is going to land soon with #5743.

You may want to hold work on this PR until then.

Sorry for the inconvenience! Thanks a lot for your work 💪

File: pytorch_lightning/cluster_environments/lsf_environment.py (hunk 3 above)
Comment (TheWitness) [reply pytorch_lightning/cluster_environments/lsf_environment.py]: The Coral deployment is a deployment on IBM Power systems that include JSM "Job Step Manager".  To get rank and size, you can use:

`LSB_MCPU_HOSTS` which follow as such "hostnameA num_ranks hostnameB num_ranks ..."
`LSB_DJOB_RANKFIKLE` which includes and OpenMPI compatible rankfile format stating specifically which cores/threads to bind to on reach host.

Of course the GPU's to use are in both the CUDA and NVIDIA environment variables.  Inside a standard job, one contained in a control group, you can only see the GPU's that you requested, and they all start with GPU0, so regardless of the actual GPU's you have allocated, they start 0, 1, 2, ... inside the control group.

**Comment by ajtritt** [comment]
@carmocca no worries, thanks for the heads up. I'll hold off for now.

**Comment by edenlightning** [comment]
@ajtritt refactors are now done! Want to take another stab at this?

**Comment by carmocca** [comment]
Note that https://github.com/PyTorchLightning/pytorch-lightning/pull/5915 is another in-progress PR which will conflict with this

**Comment by ajtritt** [comment]
@carmocca @edenlightning It's not clear to me that now is a good time to come back to this. Should I just wait until #5915 is done?

**Comment by awaelchli** [comment]
Hi, actually I wasn't aware of this PR, sorry about that. But I'm happy to assist with rebasing and otherwise if you like. The main goal of #5915 is to standardize internally how we access and interact with the cluster environment. Would love your thoughts on the interface there then this would hopefully make it much easier to support the LSF format :)

File: pytorch_lightning/plugins/environments/slurm_environment.py [hunk 5]
```diff
@@ -20,64 +20,154 @@
...
-        os.environ["MASTER_ADDR"] = root_node
+    SLURM_LOCALID
+      The node local rank for the task.
+
+    SLURM_PROCID
+      The MPI rank or relative process ID
+
+    SLURM_STEP_NUM_TASKS
+      The world size for the job. This environment variable is set by srun
+    """
```
Comment (awaelchli) [inline pytorch_lightning/plugins/environments/slurm_environment.py:44]: nice docs! 
It would be awesome if you could break this out into a separate PR :)

Comment (ajtritt) [reply pytorch_lightning/plugins/environments/slurm_environment.py:44]: Yeah, I'm happy to separate this.

**Comment by ajtritt** [comment]
@awaelchli How much does #5915 overlap with this PR?

**Comment by awaelchli** [comment]
I had a quick glance and it looks like the overlap is very minimal. You made some changes to the torchelastic and slurm environments but these could be separated from this PR, discussend and possibly be a follow up to #5915 for example.

**Comment by ajtritt** [comment]
Since #5915 adds a base cluster environment class and most of what I've done to existing code was clean things up, I think it would be easier to wait for the base class

**Comment by awaelchli** [comment]
Let's add the LSF environment class here, I'm going to add tests too. 
@ajtritt also added docs improvements to the other plugins, but I removed them from this PR. 
We can pick them up again from before my commit 
d17281c and add them in a separate PR.

**Review by awaelchli — APPROVED** [review]

**Comment by awaelchli** [comment]
@ajtritt I implemented the tests. Could you check my edits and my previous comment? Would be great to have this in 1.4. I will request review from others.

File: pytorch_lightning/plugins/environments/lsf_environment.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+    def __init__(self):
+        self._master_address = self._get_master_address()
+        self._master_port = self._get_master_port()
+        log.debug(f"MASTER_ADDR: {self._master_address}")
+        log.debug(f"MASTER_PORT: {self._master_port}")
+
+    @staticmethod
+    def is_using_lsf() -> bool:
+        """ Returns ``True`` if the current process was launched using the jsrun command. """
+        required_env_vars = (
```
Comment (tchaton) [inline pytorch_lightning/plugins/environments/lsf_environment.py:53]: Should we make them static attribute ?

File: pytorch_lightning/plugins/environments/lsf_environment.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+        """ The master address is read from a list of hosts contained in the environment variable `LSB_HOSTS`. """
+        return self._master_address
+
+    def master_port(self):
+        """ THe master port gets calculated from the LSF job ID. """
+        return self._master_port
+
+    def world_size(self):
+        """ The world size is read from the environment variable `JSM_NAMESPACE_SIZE`. """
+        var = "JSM_NAMESPACE_SIZE"
```
Comment (tchaton) [inline pytorch_lightning/plugins/environments/lsf_environment.py:74]: Same here ? To avoid possible typos ?

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/plugins/environments/lsf_environment.py (hunk 7 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/environments/lsf_environment.py:74]: yes, this would apply to all environments because it's part of the base interface. 
This todo is tracked here: #6303
Could be a good issue for a new contributor.

File: pytorch_lightning/plugins/environments/lsf_environment.py (hunk 6 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/environments/lsf_environment.py:53]: see comment below <3

**Review by ethanwharris — APPROVED** [review]
LGTM :smiley:
